### PR TITLE
Write VCFs in Python

### PR DESCRIPTION
**Pull request by jeromekelleher**
There's lots of problems with our VCF output, and fixing them is really tedious when output is done in C. This changes to using numpy tricks to generate the VCF rather than low-level C. Hopefully we don't lose too much perf (will update with some numbers).

I'd propose we get rid of VCF support in the C library entirely then.

This is surely a win, right?

### CODE REVIEW

**Comment by petrelharp** [comment]
> This is surely a win, right?

Yes!!!

**Comment by jeromekelleher** [comment]
There is a significant perf hit in the current version: we've gone from 2.5 minutes to 10 minutes for a 100K sample tree sequence. It looks like most of the time is spent in unicode encoding though, which shouldn't be necessary.

**Comment by jeromekelleher** [comment]
Latest updates bring this down to 4 minutes.  I'm willing to take a less than 2X hit here --- does it really matter if VCF output is slightly slower? It's still outputting at around 350M/s, which is as fast as most storage will be able to read it. Plus, I think a lot of the time is being spent in the header --- it's pretty stupid right now.

File: python/tskit/vcf.py [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+    n = ts.get_sample_size() // ploidy
+    sample_names = ["msp_{}".format(j) for j in range(n)]
+    last_pos = 0
+    positions = []
+    for variant in ts.variants():
+        pos = int(round(variant.position))
+        if pos <= last_pos:
+            pos = last_pos + 1
+        positions.append(pos)
+        last_pos = pos
```
Comment (petrelharp) [inline python/tskit/vcf.py]: Here's a much faster way to do this bit:
```
positions = np.round(ts.tables.sites.position)
positions[1:] += np.cumsum(np.diff(positions) == 0)
```

Comment (jeromekelleher) [reply python/tskit/vcf.py]: Nice, thanks @petrelharp.

**Comment by jeromekelleher** [comment]
TODO: rerun the perf check above, we were generating all **variants** in the header, which would massively slow things down. So, we should be more-or-less the same as the old C code now.

**Comment by jeromekelleher** [comment]
> TODO: rerun the perf check above, we were generating all variants in the header, which would massively slow things down. So, we should be more-or-less the same as the old C code now.

It's down to 3.25 minutes now, of which ~3 seconds are spent in the header. So that's fast enough; job done. I'll update with some sane default behaviour and basic support for individuals soon.

**Comment by jeromekelleher** [comment]
I think this is basically done except for a few more tests (need to check the location_transform and individual_names inputs) and to write some documentation. What do you think @petrelharp? Do you like the ``location_transform`` option? I think we need this flexibility, as there's the uber-bioinformatics confusion about whether we're using 1 or 0 based coordinates in VCF and with real data we're going to need to support both probably. The default policy is to just leave the coordinates as they are (which is what you want for simulated data, presumably).

I've thought a bit about how to include metadata as INFO fields, but I think we'll have to wait for proper metadata decoding within tskit for that. At least this does sane things site positions by default (rounds them), puts in the real allele values and uses individuals if they are present. This is a *big* improvement on what's currently available.

File: python/tskit/vcf.py (hunk 1 above)
Comment (petrelharp) [reply python/tskit/vcf.py]: Whoops; I see that my proposal didn't actually do what I wanted. I see you aren't using it anyhow... but here's code that actually does the same thing as the `legacy_position_transform`:
```
def position_transform(positions, max_iter = 100):
    positions = np.round(positions).astype("int")
    for _ in range(max_iter):
        a = np.where(np.diff(positions) == 0)[0] + 1
        if len(a) == 0:
            break
        positions[a] += 1

# test it
def gen_positions(n, m):
    out = m * np.random.random(n)
    out.sort()
    return out

for _ in range(100):
    p1 = gen_positions(10, 20)
    p2 = p1.copy()
    transform(p1)
    legacy_position_transform(p2)
    assert(np.all(p1 == p2))
```

**Comment by petrelharp** [comment]
>  Do you like the location_transform option? I think we need this flexibility, as there's the uber-bioinformatics confusion about whether we're using 1 or 0 based coordinates in VCF and with real data we're going to need to support both probably. The default policy is to just leave the coordinates as they are (which is what you want for simulated data, presumably).

Yes, I like it (including the default leave-as-is option!).  I provided some actually-correct numpy code for the "legacy" version.

**Comment by petrelharp** [comment]
Looks good to me. Thanks for getting to this!

**Comment by jeromekelleher** [comment]
> Yes, I like it (including the default leave-as-is option!). I provided some actually-correct numpy code for the "legacy" version.

Great, I'll finish this up and merge it so when I get a chance.

File: python/tskit/vcf.py (hunk 1 above)
Comment (jeromekelleher) [reply python/tskit/vcf.py]: Hmm, that still doesn't do exactly the same thing. It doesn't matter anyway --- this is legacy behaviour, and for large n the time spent doing this is trivial.

**Comment by jeromekelleher** [comment]
Closes #2 
Closes #30 
Closes  #73

**Comment by jeromekelleher** [comment]
OK, going to merge this.

**Pull request merged**